### DISCUSSION TITLE
Unexpected msg.sender and Owner Equality in Solidity Testing with Forge

### DISCUSSION THREAD

**Discussion by JarvisChan666** [Cyfrin/foundry-full-course-cu#1659] [Q&A]
## Summary
I'm encountering unexpected behavior when testing a Solidity contract suite using Foundry's Forge testing framework. Specifically, in a unit test, an assertion that checks if the contract owner **(fundMe.owner())** is equal to **deploy script** unexpectedly don't passes, where logically, it should pass, given the deployment process.

## Contracts Overview:
FundMe: Allows users to fund the contract with ETH based on USD value, utilizing a Chainlink Price Feed.
DeployFundMe: A script that deploys the FundMe contract.
FundMeTest: Contains various tests, including an assertion that fundMe.owner() should equal msg.sender.

## Problem:
In the FundMeTest, a specific test (test1()) asserts that fundMe owner should be the guy who deploy it. Given the deployment sequence, wherein DeployFundMe deploys FundMe, making itself (DeployFundMe) the owner, it's unexpected for the FundMeTest contract (acting as msg.sender) to be recognized as the FundMe contract's owner(when i use msg.sender instead of address(deployFundMe), it passed, which mean the owner of fundMe is test contract).

Code Snippets:
```solidity
// FundMe.sol
contract FundMe {
  constructor(address priceFeedAddress) {
          // If we just deploy FundMe,
          // owner will be our own address
          owner = msg.sender;
          priceFeed = AggregatorV3Interface(priceFeedAddress);
      }
}


// DeployFundMe.s.sol
contract DeployFundMe is Script {
    function run() external returns (FundMe) {
        HelperConfig helperConfig = new HelperConfig();
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        return fundMe;
    }
}

// FundMeTest.sol
contract FundMeTest is Test {
    FundMe private fundMe;
    DeployFundMe private deployFundMe;
    address private FUNDER = makeAddr("jarvis");
    uint256 private fundAmount = 1e18;

    /// @dev Sets up a new instance of the FundMe contract for each test case.
    /// Fundme owner is deploy script
    function setUp() external {
        deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function test1() public {
        assertEq(fundMe.owner(), address(deployFundMe));
    }
```

**Comment by EngrPips** [comment]
Hello @JarvisChan666, The `DeploFundMe` script wouldn't be the owner of the `fundMe` contract because you wrap the line of code that deployed the `fundMe` contract between a `vm.startBroadcast()` and a `vm.stopBroadcast()` cheat code. Like shown below 

```javascript
        vm.startBroadcast();
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
```
Whenever you use the `vm.startBroadcast()` cheat code and you don't pass in any private key, Anvil automatically has a default private key it uses, which is associated with a specific address. That address would be the sender of whatever transaction is sent between the `vm.startBroadcast()` and `vm.stopBroadcast()` cheat code. You can use console.log(), the owner of the `fundMe` contract, and console.log(), the `DeploFundMe` contract address. You will see they are different

**Comment by EngrPips** [comment]
You can either remove the `vm.startBroadcast()` and `vm.stopBroadcast()` cheat codes and then rerun the test, which should pass, or get the address that is the default sender for the foundry broadcast transaction and assert the owner of the `fundMe` contract against it.

**Reply by JarvisChan666** [reply]
Thx, problem solved

**Reply by EngrPips** [reply]
Excellent, Keep crushing it!
